### ISSUE TITLE
Express, Webpack2, RHL3 setup doesn't work

### ISSUE THREAD

**Issue by zhenyulin** [gaearon/react-hot-loader#580]
package.json:
```
"react-router": "^3.0.0",
    "react-router-redux": "^4.0.7",
"react-hot-loader": "3.0.0-beta.6",
     "webpack": "^2.6.1",
    "webpack-dev-middleware-webpack-2": "^1.5.1",
    "webpack-dev-server": "^2.4.5",
    "webpack-hot-middleware": "^2.18.0",
    "webpack-module-hot-accept": "^1.0.5"
```

.babelrc
```
{
  "plugins": [
    "transform-class-properties",
    "transform-runtime",
    "add-module-exports",
    "transform-html-import-to-string",
    "react-hot-loader/babel"
  ],
  "presets": [
    [
      "env", {
        "modules": false
      }
    ],
    "latest",
    "react",
    "flow",
  ],
}
```

Express server.js:
```
if (process.env.NODE_ENV !== 'production') {
  const webpack = require('webpack');
  const webpackDevMiddleware = require('webpack-dev-middleware-webpack-2');
  const webpackHotMiddleware = require('webpack-hot-middleware');
  const webpackConfig = require('config/webpack.dev');
  const compiler = webpack(webpackConfig);
  app.use(webpackDevMiddleware(compiler, {
    noInfo: true,
    publicPath: webpackConfig.output.publicPath,
  }));
  app.use(webpackHotMiddleware(compiler));
} 
```

webpack.dev.js
```
import path from 'path';

import webpack from 'webpack';
import HTMLWebpackPlugin from 'html-webpack-plugin';

process.env.NODE_ENV = process.env.NODE_ENV || 'development';

export default {
  devtool: 'inline-source-map',
  entry: [
    'react-hot-loader/patch',
    'webpack-hot-middleware/client',
    path.resolve('./client/index.js'),
  ],
  output: {
    path: path.resolve('./client/'),
    filename: 'bundle.js',
    publicPath: '/',
  },
  resolve: {
    modules: [
    // NOTE: new resolve pathes need to be above node_modules
      path.resolve('./client'),
      'node_modules',
    ],
    extensions: ['.js', '.jsx'],
    alias: {
      request: 'browser-request',
    },
  },
  plugins: [
    new webpack.HotModuleReplacementPlugin(),
    new webpack.NamedModulesPlugin(),
    new webpack.NoEmitOnErrorsPlugin(),
    new HTMLWebpackPlugin({
      filename: 'index.html',
      template: './client/index.html',
    }),
  ],
  module: {
    rules: [
      {
        test: /\.js$/,
        use: ['babel-loader'],
        exclude: /node_modules/,
      },
      {
        test: /\.(png|jpg|jpeg|gif|woff|woff2|svg|eot|ttf|otf|wav|mp3)$/,
        use: [
          {
            loader: 'file-loader',
            options: {
              name: '[path][name]_[hash:base64:5].[ext]',
            },
          },
        ],
      },
    ],
  },
};
```

client/index.js
```
import React from 'react';
import ReactDOM from 'react-dom';
import { AppContainer } from 'react-hot-loader';

import App from './App';

const render = (Component) => {
  ReactDOM.render(
    <AppContainer>
      <Component />
    </AppContainer>,
    document.getElementById('app'),
  );
};

render(App);

if (module.hot) {
  module.hot.accept('./App', () => { render(App); });
}

```

error message when load the page
```
[HMR] connected
(index):1 GET http://localhost:3000/__webpack_hmr net::ERR_INCOMPLETE_CHUNKED_ENCODING

```
error message in some other occasions
```
Uncaught TypeError: Cannot read property 'replaceChild' of null
```

and no update when file changes on frontend

**Comment by chrisblossom** [comment]
I think the issue is that you are using ``babel-preset-latest`` with ``babel-preset-env``. Remove latest and see what happens. Regardless, I'm pretty sure you don't want to use those presets together.

**Comment by zhenyulin** [comment]
@chrisblossom tried remove the `babel-preset-latest`, it still has the same problem.

**Comment by chrisblossom** [comment]
Can you put up a minimal example repo?

**Comment by zhenyulin** [comment]
@chrisblossom I created the problem while upgrading Webpack and React Hot Loader for an experimental boilerplate repo I have. You can find the current situation under the `dev` branch of [Hangman](https://github.com/zhenyulin/Hangman). Thank you very much.

**Comment by chrisblossom** [comment]
HMR fixed in PR https://github.com/zhenyulin/hangman/pull/3

Could be of help for others as well if they look at the commits.

**Comment by zhenyulin** [comment]
HMR fixed in another PR https://github.com/zhenyulin/hangman/pull/2

There're majorly two problems:
1) `babel-preset-flow` doesn't work well with `react-hot-loader`, which consequently breaks the HMR connection (because of some code got stripped away)
2) disable module feature in babel-preset may constraint some server-side babel-node behaviour, so probably it is better to do `const HotApp = require('./App).default` on the client side

**Comment by chrisblossom** [comment]
@zhenyulin I think you misunderstand what was causing your HMR issues. If you'd like, I can go though all the commits I made with you on Discord or Skype chat. On your fixed PR you [happened to copy the essential](https://github.com/zhenyulin/hangman/pull/2/commits/879141edaea65f3431f1a94d1bd09fe0a2eb8e6b) fixes from my PR that were causing the issue. One major being your ``nodemon`` setup.

For others having issues, ``babel-preset-flow`` does not have any issues with HMR that I am aware of..I use this in my projects without issue [(which is included by react by default)](https://github.com/babel/babel/blob/master/packages/babel-preset-react/src/index.js#L1). Also, modules work as expected as described [here](https://github.com/gaearon/react-hot-loader/tree/master/docs#webpack-2).

**Comment by zhenyulin** [comment]
@chrisblossom thanks for this. Undoubtedly, you've provided a very good solution towards the problem, which helps me to find out the minimum change I need to make to fix the issue. Further `babel-preset-env` used in your solution is a more robust preset, as `babel-preset-latest` is being deprecated.

I hope my PR here will help others upgrade to Webpack 2 and React-hot-loader 3 on express, in a easier way. In fact, I've created a new repo to recreate the status with minimum code. https://github.com/zhenyulin/minimal-express-webpack2-rhl3

Hope this helps more who are upgrading.

**Comment by zhenyulin** [comment]
@chrisblossom tested in the recreated env `babel-preset-flow` isn't causing the problem, it might be with something else together on the previous repo

**Comment by chrisblossom** [comment]
@zachasme Glad you got everything working! Good idea to close this issue if everything is fixed?

**Comment by zhenyulin** [comment]
It turns out that you don't actually need `react-hot-loader`, webpack-hmr works for most of the cases

**Comment by zhenyulin** [comment]
https://hackernoon.com/hot-reload-all-the-things-ec0fed8ab0

**Comment by amannn** [comment]
I had an issue with the handling of runtime errors with `webpack-hot-middleware`. The overlay was never shown and the app didn't recover.

I noticed that `webpack-dev-server` does a full reload when an error happens. I [added that to index.js](https://github.com/amannn/react-hot-boilerplate/commit/c4a642bfe5e20453d0ad7c2a54091b2302f1a9c2#diff-1fdf421c05c1140f6d71444ea2b27638) in order to get that working:

```js
if (module.hot) module.hot.accept('./App', () => {
  try {
    render(App)
  } catch (e) {
    location.reload();
  }
});
```

I also tried the `reload` option of `webpack-hot-middleware` but that didn't work for me.

I'm using the currently latest versions of all dependencies. Maybe this is helpful to somebody who has the same issue. Or did somebody get this working without the `try / catch`?
